File: sdk-ts/src/transport.ts
// transport.ts builds the broker Connect client from the worker's address + TLS config —
// the plumbing the SDK hides from a provider author: they set BROKER_ADDR (and, for prod,
// TLS_CERT_FILE/TLS_KEY_FILE/BROKER_CA_FILE) and never touch Connect, the h2c handshake, the
// HTTP/2 session, or cert hot-reload.
//
// The bidi WorkStream requires end-to-end HTTP/2, so the transport is always httpVersion
// "2": over a plain http:// baseUrl that is prior-knowledge cleartext h2c (NOT HTTP/1.1,
// which the broker rejects); over https:// it is HTTP/2 with the worker's mTLS material.
import { readFileSync } from "node:fs";
import { createGrpcTransport, compressionGzip } from "@connectrpc/connect-node";
import { createClient, type Client, type Transport as ConnectTransport } from "@connectrpc/connect";
import { WorkerService } from "../gen/converge/worker/v1/worker_pb.js";
import type { Config, Logger } from "./config.js";

// MaxMessageBytes bounds a single Connect message AFTER decompression — the SAME cap the
// broker enforces (pkg/wirelimits.MaxMessageBytes = 512 MiB) so a large-but-legitimate
// spec/subtree isn't rejected while a pathologically large one is bounded. Kept in sync
// with the Go const by hand (both mirror the shared wirelimits value).
export const MaxMessageBytes = 512 * 1024 * 1024;

// WorkerClient is the typed Connect client for the WorkerService — the bidi WorkStream +
// unary GetProviderConfig. It is the concrete shape the run loop + config cache speak; the
// low-level runWorker tier accepts it as its injected `client` (see the Transport alias in
// runworker.ts).
export type WorkerClient = Client<typeof WorkerService>;

// dial builds the WorkerService Connect client from cfg.brokerAddr + the TLS options. It
// reads the mTLS material once here; for cert HOT-RELOAD it re-reads on a timer via watchTLS
// (started by the caller when TLS is set), rebuilding the client so a rotated cert/CA takes
// effect with no restart. gzip compresses outbound messages (a large COMPOSE result gzips to
// a fraction); the MaxBytes caps bound the DECOMPRESSED size on both directions.
export function dial(cfg: Config): WorkerClient {
  const transport = buildTransport(cfg);
  return createClient(WorkerService, transport);
}

// buildTransport builds the underlying Connect gRPC transport. The gRPC transport is ALWAYS
// HTTP/2 (createGrpcTransport needs no httpVersion discriminator — the bidi WorkStream
// requires end-to-end HTTP/2): a cleartext http:// baseUrl is prior-knowledge h2c (no TLS);
// an https:// baseUrl with TLS_CERT_FILE/TLS_KEY_FILE set is mTLS (present the client cert
// via nodeOptions, verify the broker against BROKER_CA_FILE). HTTP/2 multiplexes the one
// long-lived WorkStream and the unary GetProviderConfig calls over a single connection — no
// HTTP/1.1 connection-storm to tune (unlike Go, whose HTTP/1.1 fallback raises
// MaxIdleConnsPerHost).
function buildTransport(cfg: Config): ConnectTransport {
  const nodeOptions = tlsNodeOptions(cfg);
  return createGrpcTransport({
    baseUrl: cfg.brokerAddr,
    // gzip OUTBOUND (compressMinBytes leaves tiny frames uncompressed); the client also
    // advertises Accept-Encoding: gzip by default, so the broker's replies are compressed too.
    sendCompression: compressionGzip,
    acceptCompression: [compressionGzip],
    readMaxBytes: MaxMessageBytes,
    writeMaxBytes: MaxMessageBytes,
    ...(nodeOptions ? { nodeOptions } : {}),
  });
}

// tlsNodeOptions returns the HTTP/2 session TLS material, or undefined for the cleartext
// h2c path. When TLS_CERT_FILE + TLS_KEY_FILE are set it loads the worker's client keypair
// (mTLS) and, if BROKER_CA_FILE is set, the CA to verify the broker; without a CA it trusts
// the system roots (unchanged). The files are read fresh on each call, so a caller that
// rebuilds the transport on a reload tick (watchTLS) picks up a rotated cert/CA.
function tlsNodeOptions(cfg: Config): { ca?: Buffer; cert?: Buffer; key?: Buffer } | undefined {
  if (cfg.tlsCertFile === "" || cfg.tlsKeyFile === "") {
    return undefined; // h2c cleartext (the WorkStream still rides HTTP/2, just unencrypted)
  }
  const opts: { ca?: Buffer; cert?: Buffer; key?: Buffer } = {
    cert: readFileSync(cfg.tlsCertFile),
    key: readFileSync(cfg.tlsKeyFile),
  };
  if (cfg.tlsCAFile !== "") {
    opts.ca = readFileSync(cfg.tlsCAFile);
  }
  return opts;
}

// watchTLS polls the cert/key/CA files every cfg.tlsReloadIntervalMs and, when their
// contents change, invokes rebuild so the caller can swap in a freshly-dialed client. Node's
// http2 session caches its TLS context, so a rotation is applied by rebuilding the transport
// rather than re-reading per handshake; the run loop's next reconnect uses the new client.
// No-op (returns immediately) on the cleartext path. Stops when signal aborts.
export function watchTLS(cfg: Config, log: Logger, rebuild: () => void, signal: AbortSignal): void {
  if (cfg.tlsCertFile === "" || cfg.tlsKeyFile === "") return;
  let last = fingerprint(cfg);
  const timer = setInterval(() => {
    let cur: string;
    try {
      cur = fingerprint(cfg);
    } catch (err) {
      log.warn("converge: worker TLS reload failed; keeping last-good material", { error: String(err) });
      return;
    }
    if (cur !== last) {
      last = cur;
      log.info("converge: worker reloaded TLS material", { cert_file: cfg.tlsCertFile, key_file: cfg.tlsKeyFile, broker_ca: cfg.tlsCAFile });
      rebuild();
    }
  }, cfg.tlsReloadIntervalMs);
  signal.addEventListener("abort", () => clearInterval(timer), { once: true });
}

// fingerprint is the concatenated cert+key+CA bytes' length+head — a cheap change detector
// (a real rotation changes the bytes). Reading the files here surfaces a mid-rotation
// missing file as a throw the caller logs (keeping last-good).
function fingerprint(cfg: Config): string {
  const parts = [readFileSync(cfg.tlsCertFile), readFileSync(cfg.tlsKeyFile)];
  if (cfg.tlsCAFile !== "") parts.push(readFileSync(cfg.tlsCAFile));
  return parts.map((b) => `${b.length}:${b.subarray(0, 32).toString("hex")}`).join("|");
}

File: sdk-go/converge/transport.go
package converge

import (
	"context"
	"crypto/tls"
	"fmt"
	"net"
	"net/http"
	"time"

	"connectrpc.com/connect"

	"github.com/salesforce/converge/pkg/tlsreload"
	"github.com/salesforce/converge/pkg/wirelimits"
	"github.com/salesforce/converge/sdk-go/workerpb/workerpbconnect"
)

// transport.go builds the broker Connect client from the Worker's address + TLS
// options. This is the plumbing the CLIENT SDK hides from a provider author: they pass
// a broker address (and, for prod, withTLS files) and never touch connectrpc, the h2c
// handshake, the HTTP transport pool, or cert hot-reload.

// withTransport is the ADVANCED escape hatch: supply an already-built broker transport
// (the generated workerpbconnect client, or any implementation) instead of letting the
// worker dial from an address + TLS. The worker then skips its own dialing entirely and
// uses this client. For custom transports — a test's in-process fake broker, a bespoke
// HTTP client, an alternate auth scheme — mirroring the WithHTTPClient/endpoint escape
// hatches cloud SDKs expose. Most workers never need it: the env (BROKER_ADDR + TLS_*) is
// the normal path.
func withTransport(client transport) option {
	return func(c *config) { c.client = client }
}

// dial builds the broker Connect client from cfg.brokerAddr + the TLS options. It
// starts the cert-reload watcher (bound to ctx) when TLS is set. Called once, on the first
// run — so construction can't fail and the dev sees no transport types.
func (w *worker) dial(ctx context.Context) (transport, error) {
	httpClient, err := w.brokerHTTPClient(ctx)
	if err != nil {
		return nil, err
	}
	// WithSendGzip compresses the worker's OUTBOUND messages — the big one being a
	// COMPOSE result Complete (a large fan-out gzips to a small fraction). The broker
	// gzips its stream replies too (the client advertises Accept-Encoding: gzip by
	// default), so both directions are compressed; the MaxBytes caps bound the
	// DECOMPRESSED size.
	return workerpbconnect.NewWorkerServiceClient(httpClient, w.cfg.brokerAddr,
		connect.WithSendGzip(),
		connect.WithReadMaxBytes(wirelimits.MaxMessageBytes),
		connect.WithSendMaxBytes(wirelimits.MaxMessageBytes)), nil
}

// brokerHTTPClient builds the HTTP client the broker is dialed over. Without withTLS
// it's prior-knowledge cleartext h2c; with withTLS it's mTLS with hot-reloaded client
// cert + broker CA.
//
// The transport is POOL-TUNED (not http.DefaultTransport). A worker holds ONE
// long-lived bidi WorkStream AND makes unary GetProviderConfig calls (boot, every
// reconnect re-pull, the periodic refresh). On HTTP/1.1 the default
// MaxIdleConnsPerHost=2 can't reuse a warm connection under a correlated burst (a
// broker restart reconnecting 50 workers at once); raising the per-host budget lets
// them reuse a warm pool instead of storming fresh connects, bounded so a worker can't
// exhaust the broker's fds.
func (w *worker) brokerHTTPClient(ctx context.Context) (connect.HTTPClient, error) {
	tr := &http.Transport{
		Proxy:                 http.ProxyFromEnvironment,
		ForceAttemptHTTP2:     true, // multiplex over one conn when the broker offers h2 (TLS path)
		MaxIdleConns:          512,
		MaxIdleConnsPerHost:   256, // default is 2 — the connection-storm fix on HTTP/1.1
		MaxConnsPerHost:       512, // cap so a worker can't exhaust the broker's fds
		IdleConnTimeout:       90 * time.Second,
		TLSHandshakeTimeout:   10 * time.Second,
		ExpectContinueTimeout: time.Second,
		// HTTP/1.1 backstop: OS TCP keepalive probes a half-open connection so a
		// silently-partitioned worker's WorkStream read eventually errors instead of
		// blocking forever. The h2 ping is the fast path on the TLS/prod route.
		DialContext: (&net.Dialer{Timeout: 10 * time.Second, KeepAlive: 30 * time.Second}).DialContext,
	}
	if w.cfg.tlsCertFile == "" || w.cfg.tlsKeyFile == "" {
		// Cleartext path: PRIOR-KNOWLEDGE cleartext HTTP/2 (h2c), NOT HTTP/1.1. The
		// WorkStream RPC is a BIDIRECTIONAL stream, which requires end-to-end HTTP/2 — a
		// plain HTTP/1.1 connection makes the broker reject it ("505 HTTP Version Not
		// Supported"). ForceAttemptHTTP2 only upgrades over TLS (ALPN); on cleartext we
		// must ask for h2c explicitly via the stdlib Protocols. HTTP/1 is left off so
		// every dial is h2c.
		var protos http.Protocols
		protos.SetUnencryptedHTTP2(true)
		tr.Protocols = &protos
		tr.ForceAttemptHTTP2 = false // Protocols governs; avoid the h1+ForceHTTP2 ambiguity
		return &http.Client{Transport: tr}, nil
	}
	// mTLS: the worker's SHORT-LIVED client cert (and the broker CA) can be rotated
	// under a running worker (cert-manager / a SPIFFE sidecar / a mounted Secret). A
	// static Certificates slice loaded once would keep presenting the EXPIRED cert after
	// rotation and the broker's RequireAndVerifyClientCert would start rejecting it,
	// silently stranding the worker. So drive the keypair from a hot-reloading Reloader
	// via GetClientCertificate (read fresh per handshake) and re-derive RootCAs per
	// handshake too so a rotated broker CA also takes effect with no restart.
	reloader, err := tlsreload.New(w.cfg.tlsCertFile, w.cfg.tlsKeyFile, w.cfg.tlsCAFile)
	if err != nil {
		return nil, fmt.Errorf("converge: worker TLS: %w", err)
	}
	go reloader.Watch(ctx, w.cfg.tlsReloadInterval, func(changed bool, rerr error) {
		switch {
		case rerr != nil:
			w.cfg.logger.Warn("converge: worker TLS reload failed; keeping last-good material", "error", rerr)
		case changed:
			c, k, ca := reloader.Files()
			w.cfg.logger.Info("converge: worker reloaded TLS material", "cert_file", c, "key_file", k, "broker_ca", ca)
		}
	})
	tr.TLSClientConfig = &tls.Config{
		MinVersion:           tls.VersionTLS12,
		GetClientCertificate: reloader.GetClientCertificate, // hot: presented per (re)dial
	}
	// RootCAs is a plain field the transport caches, so it can't hot-reload in place.
	// Re-derive it per handshake via a DialTLSContext that clones the config and stamps
	// the CURRENT CA pool. Only when a broker CA is pinned; without it the worker trusts
	// system roots (unchanged).
	if reloader.HasCA() {
		base := tr.TLSClientConfig
		tr.DialTLSContext = func(dctx context.Context, network, addr string) (net.Conn, error) {
			cfgPerConn := base.Clone()
			cfgPerConn.RootCAs = reloader.CAPool() // fresh per dial → rotated CA picked up
			d := &tls.Dialer{Config: cfgPerConn}
			return d.DialContext(dctx, network, addr)
		}
	}
	return &http.Client{Transport: tr}, nil
}
